### PR TITLE
Redux: Add utils actions and store-based uniqueId replacement

### PR DESCRIPTION
**Pull request by artpi**
# Problem
In #1496 I needed to use lodash `uniqueId` but it keeps private increment value, so it would not survive dehydration of store. Also, it could create memory problems in SSR.
@rralian mentioned this in a comment:
https://github.com/Automattic/wp-calypso/pull/1496#discussion-diff-47968438

# Solution 
Store - based `uniqueId` and utils.
_, lodash is taken, so we can call them hyphen.js, but we dont need to.

# Testing
These are helpers, so no new functionality, but you can use code from README

### CODE REVIEW

**Comment by gziolo** [comment]
This looks great :)

File: client/state/utils/actions.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+/**
+ * Internal dependencies
+ */
+import { UTILS_UNIQUEID_INCREMENT } from 'state/action-types';
+
+export function uniqueId( prefix ) {
+	return ( dispatch, getState ) => {
+		dispatch( { type: UTILS_UNIQUEID_INCREMENT } );
+		const id = getState().utils.uniqueId + '';
```
Comment (aduth) [inline client/state/utils/actions.js]: It's a few more characters, but I think `.toString()` more clearly expresses your intent here, rather than relying on coercion (which is arguably undesirably "clever").

File: client/state/utils/README.md [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+========
+
+A module for those little helpers that need to interact woth the store.
+
+# Actions
+Used in combination with the Redux store instance `dispatch` function, actions can be used in manipulating the current global state.
+
+## uniqueId( prefix )
+Meant as a replacement for _ and lodash "uniqueId" function.
+Returns unique numeric prefixed by prefix and increments internal state, so the next uniqueId will be different.
```
Comment (aduth) [inline client/state/utils/README.md]: Action creators should return actions. Still thinking about how else we might do this...

File: client/state/index.js [hunk 3]
```diff
@@ -3,7 +3,6 @@
  */
 import thunkMiddleware from 'redux-thunk';
 import { createStore, applyMiddleware, combineReducers } from 'redux';
-
```
Comment (aduth) [inline client/state/index.js]: Can we keep this newline for consistency's sake? Not sure if this was intentional.

File: client/state/utils/README.md [hunk 4]
```diff
@@ -0,0 +1,18 @@
+Utils
+========
+
+A module for those little helpers that need to interact woth the store.
```
Comment (aduth) [inline client/state/utils/README.md]: Typo: "woth" ("with")

Comment (artpi) [inline client/state/utils/README.md]: As you probably noticed reviewing my sloppy PRs, attention to detail is not my forte :/

File: client/state/utils/README.md (hunk 2 above)
Comment (artpi) [inline client/state/utils/README.md]: They do here!
I used `thunk` here and that's how it works. It is a textbook example of `thunk` actionCreator.
You have to do `dispatch( uniqueId(  ) )`

File: client/state/index.js (hunk 3 above)
Comment (artpi) [inline client/state/index.js]: fixed

File: client/state/utils/README.md (hunk 4 above)
Comment (artpi) [inline client/state/utils/README.md]: fixed

File: client/state/utils/actions.js (hunk 1 above)
Comment (artpi) [inline client/state/utils/actions.js]: fixed

**Comment by aduth** [comment]
I see the need for a unique ID as part of the state tree, but "utils" generally doesn't seem like something that should be part of state. Maybe it's the name that's bugging me. Perhaps `misc` or `meta`?

Another thought: maybe we can create a Redux middleware that intercepts and revises an intent for unique ID?

```jsx
// state/index.js (elsewhere?)
import set from 'lodash/object/set';

const uniqueId = ( store ) => ( next ) => ( action ) => {
	if ( ! action.meta || ! action.meta.uniqueId ) {
		return next( action );
	}

	store.dispatch( incrementUniqueId() );
	set( action, action.meta.uniqueId, store.getState().meta.uniqueId );
	next( action );
};

// state/actions/notices.js
export function createNotice() {
	return {
		type: 'CREATE_NOTICE',
		notice: {
			id: null
		},
		meta: {
			uniqueId: 'notice.id'
		}
	};
}
```

**Comment by artpi** [comment]
@gziolo:
> Shouldn't it be prefix based?
Both `_` and `lodash` uniqueId are NOT prefix based, they just append. In fact, I took a line with appending straight from the _ source

**Comment by artpi** [comment]
@aduth:
In my opinion, we desperately need that in the state, because it is a part of application state.
If we dehydrate and rehydrate state we need the new ids still to be unique.

If we do it through middleware, like you propose, we still need a reducer to handle `store.dispatch( incrementUniqueId() )` and we still need to put that in state to survive rehydration.

Adding middleware can be an addition to this PR, but it wont solve the problem.

Plus, I am opposed to adding so much middleware for everything because tracing actions could become a pain.

**Comment by gziolo** [comment]
> Both _ and lodash uniqueId are NOT prefix based, they just append. In fact, I took a line with appending straight from the _ source

In that case, do we really need this prefix if id always increments? At least at this moment it looks obsolte :)

**Comment by artpi** [comment]
> In that case, do we really need this prefix if id always increments? At least at this moment it looks obsolete :)

I just threw it in to keep signature in sync with _.uniqueId and lodash.uniqueId

**Comment by aduth** [comment]
>In my opinion, we desperately need that in the state

I may not have communicated clearly, but I don't have a problem with including this in state.

I confused myself about the return value of the action creator. I think it's fine to have an action creator and reducer for incrementing the ID. Perhaps part of my confusion was the naming, since action creators should be imperative (`incrementUniqueId` rather than `uniqueId`).

I'm still a bit concerned about the naming of the folder. What else do we think might be included in `state/utils` ? Is "utility state" a thing? What are your thoughts on `meta` or `misc` ? Honestly, I'm not thrilled with any of them :smile: 

The middleware is a convenience if anything, saving the developer from concerning themselves with the incrementing and fetching from the state.

```jsx
// Before
import { incrementUniqueId } from 'state/meta/actions';

export function doThing() {
	return ( dispatch, getState ) => {
		dispatch( incrementUniqueId() );
		dispatch( {
			type: DO_THING,
			id: getState().meta.uniqueId
		} );
	};
}

// After
export function doThing() {
	return {
		type: DO_THING,
		id: null,
		meta: {
			uniqueId: 'id'
		}
	};
}
```

**Comment by artpi** [comment]
> saving the developer from concerning themselves with the incrementing and fetching from the state.
The solution I hacked through thunk returns new ID from the `dispatch`. I didnt know its possible, but apparently when actionCreator is a thunk, it can return values.
So, using my code we can:
```
import { uniqueId } from 'state/meta/actions';

export function doThing() {
    return ( dispatch, getState ) => {
        dispatch( {
            type: DO_THING,
            id: dispatch( uniqueId() )
        } );
    };
}
```
> incrementUniqueId rather than uniqueId

I really wanted to mirror underscore and lodash naming and I think its more clear what it does and what it replaces

> What are your thoughts on meta or misc? 

Not that much better than `utils` :D
I dont know how to name it better :/
`lib` ? This seems to be the name for catch - all stuff throught

File: client/state/utils/README.md (hunk 2 above)
Comment (rralian) [inline client/state/utils/README.md]: I'm confused... the `uniqueId()` thunk returns a function and not an action object, and you are passing that function as an argument into the store instance's `dispatch()` method here which [expects an action object](http://rackt.org/redux/docs/api/Store.html#dispatch). So that doesn't appear valid to me.

File: client/state/utils/actions.js [hunk 5]
```diff
@@ -0,0 +1,12 @@
+/**
+ * Internal dependencies
+ */
+import { UTILS_UNIQUEID_INCREMENT } from 'state/action-types';
+
+export function uniqueId( prefix ) {
+	return ( dispatch, getState ) => {
```
Comment (rralian) [inline client/state/utils/actions.js]: I don't understand how this is supposed to work. How are the `dispatch` and `getState` arguments getting passed into this function? I don't think they are currently, but I'm also not sure how this would be modified to work correctly.

This isn't quite an action creator... nor is it a thunk that returns an action creator, because an action is never returned and an action creator is not supposed to dispatch its own action (though a _bound action creator_ can). I wonder if it make sense to separate a pure action creator out of `uniqueId()`.

**Comment by rralian** [comment]
I'm still earning my redux sea-legs, but I can't make sense of how this would work when cross-referencing against the redux docs. Particularly in regard to this comment https://github.com/Automattic/wp-calypso/pull/1793#discussion_r48904694 (wanna make sure that doesn't get missed). To me it looks like it shouldn't work.

I think this is a case where tests would be extremely helpful, not just to ensure this is working correctly, but also to demonstrate proper usage clearly.

File: client/state/utils/actions.js (hunk 5 above)
Comment (dmsnell) [inline client/state/utils/actions.js]: @rralian if the store receives a function, the thunk middleware will run it instead of simply returning the action object. This is to allow things like asynchronous calls which shouldn't dispatch until the async part has resolved. That's why there's a `dispatch` parameter to the thunk here, because it is giving us the control to decided _when_ or _how_.

That being said, I think splitting the actual action creator would be a good idea.

```js
const responseAction = response => ( {
	type: RESPONSE_ACTION,
	response
} );

const asyncAction = dispatch => {
	getFromServer()
		.then( response => dispatch( responseAction( response ) ) )
		.catch( error.log );
};
```

File: client/state/utils/actions.js [hunk 6]
```diff
@@ -0,0 +1,12 @@
+/**
+ * Internal dependencies
+ */
+import { UTILS_UNIQUEID_INCREMENT } from 'state/action-types';
+
+export function uniqueId( prefix ) {
```
Comment (dmsnell) [inline client/state/utils/actions.js]: I'm curious why we specify `prefix` here. Is is simply a helper to make it easy to embed the counter in a string? The store clearly doesn't care about the prefix since it's never passed, meaning that `prefix` doesn't appear to have any meaning with regards to the unique id.

Is it worth adding here? Is it here just to mimic **lodash**?

File: client/state/utils/reducer.js [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+			state = state + 1;
+			break;
+	}
+
+	return state;
+}
+
+export default combineReducers( {
+	uniqueId
+} );
```
Comment (dmsnell) [inline client/state/utils/reducer.js]: We don't really seem to gain much here using `combineReducers()` except maybe for the fact that it automatically sets the key name. I can easily imagine setting up the code like this if we think we'll extend the reducers in the future, but that seems unlikely for `uniqueId`

I'm cool leaving it in, but uneasy about doing it when not needed/not adding value.

File: client/state/utils/README.md (hunk 2 above)
Comment (mcsf) [inline client/state/utils/README.md]: The rest of the `uniqueId` function looks like a regular `redux-thunk`-powered action creator, but the return value of the thunk itself is where things get funky. A thunk isn't expected to return anything at all, since it is just a delayed execution that will call `dispatch` whenever it decides.

See the [implementation of `redux-thunk`](https://github.com/gaearon/redux-thunk/blob/00f7fdb9b41e7445941692bc9c808f3769be22bd/src/index.js#L4): the output of `action( dispatch, getState )` — which is the invocation of the thunk — is never caught. I don't think it's our place to catch it.

tl;dr: a thunk should not be returning anything. I can see why this approach was followed, but by taking this shortcut — vs. going the regular Redux approach of accessing global state via `connect`+selector — we're fragmenting the way we all deal with data, and potentially opening holes through which things may fail at some point.

Comment (artpi) [inline client/state/utils/README.md]: @rralian 
> dispatch() method here which expects an action object. 

`redux-thunk` middleware introduces possibility to pass functions to dispatch. This is the whole source:
```
function thunkMiddleware({ dispatch, getState }) {
  return next => action =>
    typeof action === 'function' ?
      action(dispatch, getState) :
      next(action);
}
```
As you see, if action is a function, it gets called with `dispatch` and `getState`.

I completely agree that mechanics of this may seem funky, but I kinda feel that this is the most convenient version to use later.
On the other hand, I recently argued that readability is more important then convenience, and it seems that @dmsnell rightly pointed that readability is in the eye of the reader :)

More to the point: 

> the output of action( dispatch, getState ) — which is the invocation of the thunk — is never caught.

I let myself translate this source to ES5, and as I understand it, it gets returned:
```
function thunkMiddleware(_ref) {
  var dispatch = _ref.dispatch;
  var getState = _ref.getState;

  return function (next) {
    return function (action) {
      return typeof action === 'function' ? action(dispatch, getState) : next(action);
    };
  };
}
```

@mcsf:
> by taking this shortcut — vs. going the regular Redux approach of accessing global state via connect+selector

I took this approach to have an 'atomic' action that both increments value and returns it.
The regular redux approach would be to have action for incrementing and selector for accessing, so to get the next id, the flow would be:
1. Call the action, store value gets incremented
2. Component gets re-rendered because we connected to the store for accessing this data
3. Get new id from the props
4. Use the new id
5. call the action that needs ID
6. Store mutation, re-render...

File: client/state/utils/actions.js (hunk 6 above)
Comment (artpi) [inline client/state/utils/actions.js]: Yes, it is exactly to mimic `lodash`, it only gets returned prepended to the new id :)

File: client/state/utils/README.md (hunk 2 above)
Comment (dmsnell) [inline client/state/utils/README.md]:
```haskell
uniqueId :: IO Number
uniqueId = incrementUniqueId >> getNextUniqueId
```

:fire: :trollface:

Comment (mcsf) [inline client/state/utils/README.md]: > I let myself translate this source to ES5, and as I understand it, it gets returned:

Oh yeah, d'oh, you're absolutely right. I'd read it too fast. :)

> I took this approach to have an 'atomic' action that both increments value and returns it.
The regular redux approach would be to have action for incrementing and selector for accessing, so to get the next id, the flow would be: […]

Hm-hm, there's a lot of indirection. It's a tricky issue.

Don't get me wrong, the current implementation works and is very straightforward, and that is good. I'm being cautious because we're diverging from our standard way of thinking about global data, and I'm also wondering whether we should avoid diverging altogether **or** find a standard way to diverge. :)

In a way, the issue at hand is that we want some sort of non-component, non-UI state in our app to be Redux- (and thus, for instance, SSR-) friendly. Right? If we didn't have those concerns, a simple module-level global `_count` variable — or a closure-bound variable — would have done the job. Function memoization is also something that would fall in this category: its cache is a form of persistent/global data, and in an SSR context we'd want that data bound to each request _or_ to each React render tree (i.e. we wouldn't want it globally shared across requests).

So, where I'm getting at is that we might want a sort of standard way to bind some stateful function (`uniqueId`, `memoize`) to a React tree/context/request. Though that likely won't be solved in just this PR, it's something to keep thinking about.

Comment (mcsf) [inline client/state/utils/README.md]: > uniqueId :: IO Number
> uniqueId = incrementUniqueId >> getNextUniqueId

@dmsnell: Boom, done. Love it.

File: client/state/utils/reducer.js (hunk 7 above)
Comment (artpi) [inline client/state/utils/reducer.js]: > if we think we'll extend the reducers in the future

Exactly my intention, this file is `utils/reducer.js` and I think we may find other candidates to put here. everything thats memoizing or holding some internal state that needs to be put to the store

File: client/state/utils/README.md (hunk 2 above)
Comment (artpi) [inline client/state/utils/README.md]: @mcsf 
> here I'm getting at is that we might want a sort of standard way to bind some stateful function (uniqueId, memoize) to a React tree/context/request

Thats the whole point of this `utils` state subtree to hold this kind of functions. Maybe memoize is not the best example, because maybe we dont want to store cache in the store..

**Comment by artpi** [comment]
@rralian: 
> I can't make sense of how this would work when cross-referencing against the redux docs

Thats because `Without middleware, Redux store only supports synchronous data flow. `. I *mis*used `redux-thunk` to get store state in an action creator, I described it in detail in the comment discussion: https://github.com/Automattic/wp-calypso/pull/1793#discussion_r48969484

> I think this is a case where tests would be extremely helpful

Good point! I dont know if they are helpful, but they should be there and now they are :)

File: client/state/utils/actions.js (hunk 5 above)
Comment (artpi) [inline client/state/utils/actions.js]: @dmsnell 
> That being said, I think splitting the actual action creator would be a good idea.
Agree, done in da6dbad9d781aa5ad498dec44ca1d83116fe59e0

File: client/state/utils/test/actions.js [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+			uniqueId()( dispatchSpy, getStateSpy );
+			expect( dispatchSpy.calledWithMatch( { type: UTILS_UNIQUEID_INCREMENT } ) ).to.be.ok;
+		} );
+
+		it( 'should return new counter value', () => {
+			const dispatchSpy = spy(),
+				getStateSpy = spy( getState ),
+				newValue = uniqueId()( dispatchSpy, getStateSpy );
+
+			expect( newValue ).to.eql( '3' );
```
Comment (dmsnell) [inline client/state/utils/test/actions.js]: This test doesn't appear to actually verify that the unique id gets incremented. Could we rewrite it so that we can do that?

Comment (artpi) [inline client/state/utils/test/actions.js]: @dmsnell 
That would require importing both `actions.js` and a `reducer.js`. I dont want to do that in this file, because that would violate some benefits of unit tests (for example it would be hard to pinpoint what exactly has been broken, because tests would fail if either action file or reducer had an error)

But you're right, we should test that and I added separate 'integration' test in commit 2520c5d6d18698547ace4e9a096353c259594dd3

File: client/state/utils/actions.js [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+import { UTILS_UNIQUEID_INCREMENT } from 'state/action-types';
+
+export function uniqueIdIncrement() {
+	return { type: UTILS_UNIQUEID_INCREMENT };
+}
+
+export function uniqueId( prefix ) {
+	return ( dispatch, getState ) => {
+		dispatch( uniqueIdIncrement() );
+		const id = getState().utils.uniqueId.toString();
```
Comment (dmsnell) [inline client/state/utils/actions.js]: Why are we casting to a string?

File: client/state/utils/test/reducer.js [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+	describe( 'uniqueId', () => {
+		it( 'should default to a number', () => {
+			const state = uniqueId( undefined, {} );
+			expect( state ).to.be.a( 'number' );
+		} );
+
+		it( 'action UTILS_UNIQUEID_INCREMENT should increment previous state', () => {
+			const state = uniqueId( 1, { type: UTILS_UNIQUEID_INCREMENT } );
+			expect( state ).to.eql( 2 );
+		} );
```
Comment (dmsnell) [inline client/state/utils/test/reducer.js]: This is the test I wanted to see, thanks!

File: client/state/utils/test/integration.js [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+			const initState = 1;
+			state = initState;
+
+			const returnValue = uniqueId()( dispatchSpy, getStateSpy );
+
+			expect( state ).to.eql( initState + 1 );
+			expect( returnValue ).to.eql( state.toString() );
+		} );
+	} );
+} );
```
Comment (dmsnell) [inline client/state/utils/test/integration.js]: I love the test in `test/reducer.js` and it appears to cover the behavior of the unique id. This "integration" test also does its job but there's no real distinction between it and the other unit tests because it runs every time with them. It's also not particularly slower then them anyway.

So since we know from the unit test that the `uniqueId()` function works as expected, I don't think we need this test too. It's mainly actually testing that redux does its job, which is covered by numerous redux unit tests.

**Comment by dmsnell** [comment]
@artpi thanks for the new tests! I have a question per the need of the integration test.

File: client/state/utils/actions.js (hunk 9 above)
Comment (artpi) [inline client/state/utils/actions.js]: Because _ and lodash do so, I wanted full parity.

File: client/state/utils/test/reducer.js (hunk 10 above)
Comment (artpi) [inline client/state/utils/test/reducer.js]: It's good it's here :)

File: client/state/utils/test/integration.js (hunk 11 above)
Comment (artpi) [inline client/state/utils/test/integration.js]: Yes, that is true, but since it's already here, I don't see any harm in leaving it, to be honest.

**Comment by artpi** [comment]
Rebased.

**Comment by gwwar** [comment]
To take a step back, what's the use case of having an auto-incrementing id over a uuid?

**Comment by dmsnell** [comment]
@gwwar I don't think there'd be a problem with generating a UUID, but there's probably also not the need. The unique id also gives us a unique number that's guaranteed not to collide. As far as state goes, I think that this would still be the case with the UUID. The way I understand this, @artpi can correct, is that these are currently being used for local-only objects which might potentially be stored in a cache. In the UUID case, we'd probably still need to remember which ids we assigned to the objects when we load them back from the cache.

**Comment by gwwar** [comment]
It feels a bit odd here to store state for a utility method in the render tree.

It sounds like the current use case is to generate unique keys for an object, where the order of magnitude is around the 10s or 100s for notices. ( Let me know if had anything else in mind @artpi ). There is a little bit extra work to guarantee uniqueness for uuids (basically keep generating ids until you don't find an entry), but it should be too much trouble. Depending on how many ids we need we can also keep them short.

**Comment by artpi** [comment]
@gwwar 
>It sounds like the current use case is to generate unique keys for an object, where the order of magnitude is around the 10s or 100s for notices. ( Let me know if had anything else in mind @artpi ).

My main intention was to:
- Provide a helper method that can be used across the app
- Ensure that we don't have 'hidden' state (like in the _.uniqueID I used in notices. )

> There is a little bit extra work to guarantee uniqueness for uuids (basically keep generating ids until you don't find an entry), but it should be too much trouble

This could generate some huge mis-match because dismissing notice deletes the entry and you can dismiss notice from a lot of places.
So imagine that:
- We generate a notice
- put link to dismiss on timeout, close button and also tie dismissing to some action (ex. connection restored)
- One of this things dismisses the notice, deleting it
- Other notice gets generated with the id thats now free
- other action intended to dismiss the new notice not dismisses the old one.

Thats why uniqueID needs to be unique :)

**Comment by gwwar** [comment]
The likelihood of the above happening is very small and is of somewhat minor consequence, but I'll defer here if folks prefer an auto-incrementing id. ;)

**Comment by artpi** [comment]
New update:
Notices are excluded from persisted redux store ATM.
I am leaving this PR, maybe it will become useful.
For now it is not needed.

**Comment by aduth** [comment]
If revisited again in the future, this pull request will need to be updated with the merger of #3773:

- No need to specify custom `Makefile` in `client/state`
 - Instead, require your tests directly in `client/state/test/index.js`

**Comment by artpi** [comment]
I am closing this PR for now. If it becomes needed, I will resurect it
